### ISSUE TITLE
FM模型实现细节请教

### ISSUE THREAD

**Issue by ChengQianlong** [ZiyaoGeng/RecLearn#42]
你好，请问下FM模型51行
inputs = inputs + tf.convert_to_tensor(self.index_mapping)
为什么要给inputs加上 index_mapping 呢？index_mapping的作用是什么？
辛苦解答

**Comment by liupandatacs** [comment]
我也有这个问题，感觉不太对

**Comment by ZiyaoGeng** [comment]
是为了纪录全部特征进行拼接后的索引，申请的1阶权重参数的维度是所有特征长度之和。对于第二个特征，若想知道他在w的索引位置，需要第一个特征的长度+当前第二个特征对应的索引，才等于在w中对应的索引。
之前采用的是one-hot向量，太慢，所以采用这种方式。有些著名开源项目实现fm中包括deepfm都是把一阶直接去掉，我是采用这种方式使得模型训练速度相比one-hot更好。
如果有更好的表示方法，欢迎在这里留言

**Comment by xiyuan1223** [comment]
这里这么做是应为所有的特征都共用了 一个embedding table , 所以各个特征的index 都不能重复吧

**Comment by dxjlkj** [comment]
first_order = self.w0 + tf.reduce_sum(tf.nn.embedding_lookup(self.w, inputs), axis=1)
这段代码不是很理解，FM论文first_order中应该有w与数据x相乘的部分，这段代码是什么意思呢。
辛苦解答

**Comment by TangJiakai** [comment]
> first_order = self.w0 + tf.reduce_sum(tf.nn.embedding_lookup(self.w, inputs), axis=1) 这段代码不是很理解，FM论文first_order中应该有w与数据x相乘的部分，这段代码是什么意思呢。 辛苦解答

其他框架实现FM也是类似的，我个人是觉得因为我们都划分成类别型特征，那么w*x和对应embedding求和是等价的（个人见解）。
